include/plugin_loader.hpp

```diff
@@ -0,0 +1,196 @@
+#ifndef GRAPH_PROTOTYPE_PLUGIN_LOADER_H
+#define GRAPH_PROTOTYPE_PLUGIN_LOADER_H
+
+#include <algorithm>
+#include <filesystem>
+#include <iostream>
+#include <span>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include <dlfcn.h>
+#include <graph.hpp>
+#include <plugin.hpp>
+
+using plugin_create_function_t  = gp_plugin_base *(*) ();
+using plugin_destroy_function_t = void (*)(gp_plugin_base *);
+
+namespace fair::graph {
+
+using namespace std::string_literals;
+using namespace std::string_view_literals;
+
+#ifndef __EMSCRIPTEN__
+// Plugins are not supported on WASM
+
+class plugin_handler {
+private:
+    void                     *_dl_handle  = nullptr;
```

Comment:
It could be u_p with a custom deleter (dlclose), same for the next one.
But, that is just an indirection - plugin_handler is a safe value-type wrapper over the pair of _dl_handle/_instance pointers.